File: backend/src/main/java/com/leavemarker/service/LeavePolicyService.java
package com.leavemarker.service;

import com.leavemarker.dto.leavepolicy.LeavePolicyRequest;
import com.leavemarker.dto.leavepolicy.LeavePolicyResponse;
import com.leavemarker.entity.Company;
import com.leavemarker.entity.LeavePolicy;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.exception.ResourceNotFoundException;
import com.leavemarker.repository.CompanyRepository;
import com.leavemarker.repository.LeavePolicyRepository;
import com.leavemarker.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class LeavePolicyService {

    private final LeavePolicyRepository leavePolicyRepository;
    private final CompanyRepository companyRepository;
    private final SubscriptionFeatureService subscriptionFeatureService;

    @Transactional
    public LeavePolicyResponse createLeavePolicy(LeavePolicyRequest request, UserPrincipal currentUser) {
        Company company = companyRepository.findById(currentUser.getCompanyId())
                .orElseThrow(() -> new ResourceNotFoundException("Company not found"));

        // Check subscription plan limits
        subscriptionFeatureService.validateCanAddLeavePolicy(company.getId());

        if (leavePolicyRepository.existsByCompanyIdAndLeaveTypeAndDeletedFalse(
                company.getId(), request.getLeaveType())) {
            throw new BadRequestException("Leave policy for this leave type already exists");
        }

        if (request.getCarryForward() && (request.getMaxCarryForward() == null || request.getMaxCarryForward() < 0)) {
            throw new BadRequestException("Max carry forward must be specified when carry forward is enabled");
        }

        LeavePolicy policy = LeavePolicy.builder()
                .company(company)
                .leaveType(request.getLeaveType())
                .annualQuota(request.getAnnualQuota())
                .monthlyAccrual(request.getMonthlyAccrual())
                .carryForward(request.getCarryForward())
                .maxCarryForward(request.getMaxCarryForward() != null ? request.getMaxCarryForward() : 0)
                .encashmentAllowed(request.getEncashmentAllowed())
                .halfDayAllowed(request.getHalfDayAllowed())
                .active(request.getActive())
                .build();

        policy = leavePolicyRepository.save(policy);
        return mapToResponse(policy);
    }

    public LeavePolicyResponse getLeavePolicy(Long id, UserPrincipal currentUser) {
        LeavePolicy policy = leavePolicyRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Leave policy not found"));

        if (!policy.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        return mapToResponse(policy);
    }

    public List<LeavePolicyResponse> getAllLeavePolicies(UserPrincipal currentUser) {
        List<LeavePolicy> policies = leavePolicyRepository.findByCompanyIdAndDeletedFalse(
                currentUser.getCompanyId());
        return policies.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<LeavePolicyResponse> getActiveLeavePolicies(UserPrincipal currentUser) {
        List<LeavePolicy> policies = leavePolicyRepository.findByCompanyIdAndActiveAndDeletedFalse(
                currentUser.getCompanyId(), true);
        return policies.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public LeavePolicyResponse updateLeavePolicy(Long id, LeavePolicyRequest request, UserPrincipal currentUser) {
        LeavePolicy policy = leavePolicyRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Leave policy not found"));

        if (!policy.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        if (!policy.getLeaveType().equals(request.getLeaveType())) {
            if (leavePolicyRepository.existsByCompanyIdAndLeaveTypeAndDeletedFalse(
                    currentUser.getCompanyId(), request.getLeaveType())) {
                throw new BadRequestException("Leave policy for this leave type already exists");
            }
        }

        if (request.getCarryForward() && (request.getMaxCarryForward() == null || request.getMaxCarryForward() < 0)) {
            throw new BadRequestException("Max carry forward must be specified when carry forward is enabled");
        }

        policy.setLeaveType(request.getLeaveType());
        policy.setAnnualQuota(request.getAnnualQuota());
        policy.setMonthlyAccrual(request.getMonthlyAccrual());
        policy.setCarryForward(request.getCarryForward());
        policy.setMaxCarryForward(request.getMaxCarryForward() != null ? request.getMaxCarryForward() : 0);
        policy.setEncashmentAllowed(request.getEncashmentAllowed());
        policy.setHalfDayAllowed(request.getHalfDayAllowed());
        policy.setActive(request.getActive());

        policy = leavePolicyRepository.save(policy);
        return mapToResponse(policy);
    }

    @Transactional
    public void deleteLeavePolicy(Long id, UserPrincipal currentUser) {
        LeavePolicy policy = leavePolicyRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Leave policy not found"));

        if (!policy.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        policy.setDeleted(true);
        leavePolicyRepository.save(policy);
    }

    private LeavePolicyResponse mapToResponse(LeavePolicy policy) {
        return LeavePolicyResponse.builder()
                .id(policy.getId())
                .leaveType(policy.getLeaveType())
                .annualQuota(policy.getAnnualQuota())
                .monthlyAccrual(policy.getMonthlyAccrual())
                .carryForward(policy.getCarryForward())
                .maxCarryForward(policy.getMaxCarryForward())
                .encashmentAllowed(policy.getEncashmentAllowed())
                .halfDayAllowed(policy.getHalfDayAllowed())
                .active(policy.getActive())
                .companyId(policy.getCompany().getId())
                .companyName(policy.getCompany().getName())
                .build();
    }
}

File: backend-go/internal/service/leave_policy_service.go
package service

import (
	"github.com/leavemarker/backend-go/internal/dto"
	"github.com/leavemarker/backend-go/internal/models"
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
)

// LeavePolicyService handles leave policy business logic
type LeavePolicyService struct {
	leavePolicyRepo       *repository.LeavePolicyRepository
	planValidationService *PlanValidationService
}

// NewLeavePolicyService creates a new LeavePolicyService
func NewLeavePolicyService(
	leavePolicyRepo *repository.LeavePolicyRepository,
	planValidationService *PlanValidationService,
) *LeavePolicyService {
	return &LeavePolicyService{
		leavePolicyRepo:       leavePolicyRepo,
		planValidationService: planValidationService,
	}
}

// CreateLeavePolicy creates a new leave policy
func (s *LeavePolicyService) CreateLeavePolicy(companyID uint, req dto.LeavePolicyRequest) (*dto.LeavePolicyResponse, error) {
	// Validate plan limit
	if err := s.planValidationService.ValidateLeavePolicyLimit(companyID); err != nil {
		return nil, err
	}

	// Check if policy for this leave type already exists
	exists, err := s.leavePolicyRepo.ExistsByCompanyIDAndLeaveType(companyID, req.LeaveType)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to check leave policy")
	}
	if exists {
		return nil, utils.NewBadRequestError("Leave policy for this type already exists")
	}

	policy := &models.LeavePolicy{
		CompanyID:         companyID,
		LeaveType:         req.LeaveType,
		AnnualQuota:       req.AnnualQuota,
		MonthlyAccrual:    req.MonthlyAccrual,
		CarryForward:      req.CarryForward,
		MaxCarryForward:   req.MaxCarryForward,
		EncashmentAllowed: req.EncashmentAllowed,
		HalfDayAllowed:    req.HalfDayAllowed,
		Active:            req.Active,
	}

	if err := s.leavePolicyRepo.Create(policy); err != nil {
		return nil, utils.NewInternalServerError("Failed to create leave policy")
	}

	return s.toLeavePolicyResponse(policy), nil
}

// GetLeavePolicy gets a leave policy by ID
func (s *LeavePolicyService) GetLeavePolicy(id, companyID uint) (*dto.LeavePolicyResponse, error) {
	policy, err := s.leavePolicyRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Leave policy not found")
	}

	if policy.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	return s.toLeavePolicyResponse(policy), nil
}

// GetAllLeavePolicies gets all leave policies for a company
func (s *LeavePolicyService) GetAllLeavePolicies(companyID uint) ([]dto.LeavePolicyResponse, error) {
	policies, err := s.leavePolicyRepo.FindByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get leave policies")
	}

	responses := make([]dto.LeavePolicyResponse, len(policies))
	for i, policy := range policies {
		responses[i] = *s.toLeavePolicyResponse(&policy)
	}

	return responses, nil
}

// GetActiveLeavePolicies gets all active leave policies for a company
func (s *LeavePolicyService) GetActiveLeavePolicies(companyID uint) ([]dto.LeavePolicyResponse, error) {
	policies, err := s.leavePolicyRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get leave policies")
	}

	responses := make([]dto.LeavePolicyResponse, len(policies))
	for i, policy := range policies {
		responses[i] = *s.toLeavePolicyResponse(&policy)
	}

	return responses, nil
}

// UpdateLeavePolicy updates a leave policy
func (s *LeavePolicyService) UpdateLeavePolicy(id, companyID uint, req dto.LeavePolicyRequest) (*dto.LeavePolicyResponse, error) {
	policy, err := s.leavePolicyRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Leave policy not found")
	}

	if policy.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	// Check if changing leave type and if new type already exists
	if req.LeaveType != policy.LeaveType {
		exists, err := s.leavePolicyRepo.ExistsByCompanyIDAndLeaveType(companyID, req.LeaveType)
		if err != nil {
			return nil, utils.NewInternalServerError("Failed to check leave policy")
		}
		if exists {
			return nil, utils.NewBadRequestError("Leave policy for this type already exists")
		}
	}

	policy.LeaveType = req.LeaveType
	policy.AnnualQuota = req.AnnualQuota
	policy.MonthlyAccrual = req.MonthlyAccrual
	policy.CarryForward = req.CarryForward
	policy.MaxCarryForward = req.MaxCarryForward
	policy.EncashmentAllowed = req.EncashmentAllowed
	policy.HalfDayAllowed = req.HalfDayAllowed
	policy.Active = req.Active

	if err := s.leavePolicyRepo.Update(policy); err != nil {
		return nil, utils.NewInternalServerError("Failed to update leave policy")
	}

	return s.toLeavePolicyResponse(policy), nil
}

// DeleteLeavePolicy soft deletes a leave policy
func (s *LeavePolicyService) DeleteLeavePolicy(id, companyID uint) error {
	policy, err := s.leavePolicyRepo.FindByID(id)
	if err != nil {
		return utils.NewNotFoundError("Leave policy not found")
	}

	if policy.CompanyID != companyID {
		return utils.NewForbiddenError("Access denied")
	}

	return s.leavePolicyRepo.Delete(id)
}

func (s *LeavePolicyService) toLeavePolicyResponse(policy *models.LeavePolicy) *dto.LeavePolicyResponse {
	return &dto.LeavePolicyResponse{
		ID:                policy.ID,
		LeaveType:         policy.LeaveType,
		AnnualQuota:       policy.AnnualQuota,
		MonthlyAccrual:    policy.MonthlyAccrual,
		CarryForward:      policy.CarryForward,
		MaxCarryForward:   policy.MaxCarryForward,
		EncashmentAllowed: policy.EncashmentAllowed,
		HalfDayAllowed:    policy.HalfDayAllowed,
		Active:            policy.Active,
		CreatedAt:         policy.CreatedAt,
		UpdatedAt:         policy.UpdatedAt,
	}
}
